### PR TITLE
[RFC] improve CursorMoved

### LINKED ISSUE #9755: CursorMoved issue with aucmd_prepbuf when using nvim_buf_set_lines()
```vim
function! s:foo() abort
  let lines = ['one', 'two', 'three']
  let buf = nvim_create_buf(v:false, v:true)
  call nvim_buf_set_lines(buf, 0, -1, v:true, lines) " culprit line
  let s:win = nvim_open_win(buf, 0, {
        \ 'relative': 'win',
        \ 'anchor': 'SW',
        \ 'width': winwidth('%'),
        \ 'height': len(lines),
        \ 'col': 0,
        \ 'row': winheight('%')
        \ })
  autocmd CursorMoved <buffer> ++once call nvim_win_close(s:win, v:false)
endfunction

command! Foo call s:foo()
```

The code is supposed to open a floating window, without entering it, which gets deleted again on `CursorMoved`. But the autocmd is triggered right away, so the floating window is created and deleted in one go.

It works as expected when commenting out `nvim_buf_set_lines()`.

It's most probably related to the use of `autocmd_prefbuf()`:

```diff
diff --git i/src/nvim/api/buffer.c w/src/nvim/api/buffer.c
index 3613a8f8b..4621f61d4 100644
--- i/src/nvim/api/buffer.c
+++ w/src/nvim/api/buffer.c
@@ -396,7 +396,7 @@ void nvim_buf_set_lines(uint64_t channel_id,

   try_start();
   aco_save_T aco;
-  aucmd_prepbuf(&aco, (buf_T *)buf);
+  /* aucmd_prepbuf(&aco, (buf_T *)buf); */

   if (u_save((linenr_T)(start - 1), (linenr_T)end) == FAIL) {
     api_set_error(err, kErrorTypeException, "Failed to save undo information");
@@ -477,7 +477,7 @@ end:
   }

   xfree(lines);
-  aucmd_restbuf(&aco);
+  /* aucmd_restbuf(&aco); */
   try_end(err);
 }

```

Afterwards, the floating window appears (and stays) as expected, but the text is appended to the wrong buffer (the current one instead of the one created by `nvim_create_buf()`).

Slapping `set eventignore=CursorMoved` around `nvim_buf_set_lines()` or using `:noautocmd` has no effect.

### PR DESCRIPTION
**Pull request by mhinz**
Fixes https://github.com/neovim/neovim/issues/9755

### CODE REVIEW

File: src/nvim/buffer_defs.h [hunk 1]
```diff
@@ -1183,6 +1183,8 @@ struct window_S {
   linenr_T w_redraw_bot;            /* when != 0: last line needing redraw */
   int w_redr_status;                /* if TRUE status line must be redrawn */
 
+  pos_T w_last_cursormoved;         ///< for CursorMoved event
```
Comment (justinmk) [inline src/nvim/buffer_defs.h]: maybe it should go next to `w_last_cursorline`

**Review by justinmk — APPROVED** [review]

**Comment by andymass** [comment]
Could/should this change be upstreamed?

**Comment by clason** [comment]
I have a weird performance issue that I have bisected to this commit (this time really), involving floating windows containing omnicomplete info previews triggered in a file with heavy syntax highlighting and delimiter matching. The issue is that the (quite expensive) delimiter matching function is triggered on cursor movement (not <c-n>!) in the popupmenu even though it shouldn't (see prabirshrestha/vim-lsp#507 -- warning: long back-and-forth).

Specifically, it's the change in line 4264 in `src/nvim/window.c` -- if I comment out that line (also on master), the issue goes away. I'd be happy to open an issue about it if you think it actually is one. Any hints on useful information I could provide there would be appreciated -- reproducing this is kind of involved.

**Comment by blueyed** [comment]
This line? https://github.com/neovim/neovim/pull/9807/files#diff-40d9eb21ee75b3df148c827ec4d4310aR4263

**Comment by clason** [comment]
Yes; I'm trying to understand whether this does what it's supposed to and the problem is somewhere else.

**Comment by bfredl** [comment]
Discussed on gitter. For posterity: this was because `nvim_win_open` was called with `enter=v:true` https://github.com/prabirshrestha/vim-lsp/blob/da078b1096944d149a0593dc8e49ae097d7b4ed3/autoload/lsp/ui/vim/output.vim#L116 where autocmd firing is expected behavior. To avoid this issue, rewrite code to use `enter=v:false`.

**Pull request merged**